modules/qualitycontrol/QualityController.ts

```diff
@@ -0,0 +1,366 @@
+                .map((stat: ITrackStats) => stat.encodeTime)
+                .reduce((totalValue: number, currentValue: number) => totalValue + currentValue, 0);
+            const avgEncodeTime: number = totalEncodeTime / trackStats.length;
+            const { qualityLimitationReason = QualityLimitationReason.NONE }
+                = trackStats
+                    .find((stat: ITrackStats) => stat.qualityLimitationReason !== QualityLimitationReason.NONE) ?? {};
+            const encodeResolution: number = trackStats
+                .map((stat: ITrackStats) => stat.encodeResolution)
+                .reduce((resolution: number, currentValue: number) => Math.max(resolution, currentValue), 0);
+            const localTrack = this._conference.getLocalVideoTracks().find(t => t.rtcId === trackId);
+
+            const exisitingStats: FixedSizeArray = this._encodeTimeStats.get(trackId);
+            const sourceStats = {
+                avgEncodeTime,
+                codec,
+                encodeResolution,
+                qualityLimitationReason,
+                localTrack,
+                timestamp
+            };
+
+            if (exisitingStats) {
+                exisitingStats.add(sourceStats);
+            } else {
+                // Save stats for only the last 5 mins.
+                const data = new FixedSizeArray(300);
+
+                data.add(sourceStats);
+                this._encodeTimeStats.set(trackId, data);
+            }
+
+            logger.debug(`Encode stats for ${localTrack}: codec=${codec}, time=${avgEncodeTime},`
+                + `resolution=${encodeResolution}, qualityLimitationReason=${qualityLimitationReason}`);
+
+            // Do not attempt run time adjustments if the adaptive mode is disabled.
+            if (!this._enableAdaptiveMode) {
+                return;
+            }
+
+            if (encodeResolution === localTrack.maxEnabledResolution) {
+                if (this._limitedByCpuTimeout) {
+                    window.clearTimeout(this._limitedByCpuTimeout);
+                    this._limitedByCpuTimeout = undefined;
+                }
+
+                return;
+            }
+
+            // Do nothing if the limitation reason is bandwidth since the browser will dynamically adapt the outbound
+            // resolution based on available uplink bandwith. Otherwise,
+            // 1. Switch the codec to the lowest complexity one incrementally.
+            // 2. Switch to a lower lastN value, cutting the receive videos by half in every iteration until
+            // LAST_N_THRESHOLD is reached.
+            // 3. Lower the receive resolution of individual streams up to 180p.
+            if (qualityLimitationReason === QualityLimitationReason.CPU
+                || (encodeResolution < Math.min(localTrack.maxEnabledResolution, localTrack.getCaptureResolution())
+                    && qualityLimitationReason !== QualityLimitationReason.BANDWIDTH)) {
```

Comment:
This was to handle the case where the browser doesn't report quality limitation in the stats and yet you see a degraded outbound stream as observed on your Acer laptop. It all depends on how we tackle that case. Maybe we need more datapoints to determine that is the case. Compute pressure doesn't seem to be relevant at all. It would be great if you can test on your laptop and share your findings so that we can adjust the check here.

Reply:
Will do!